File: e22093-Refactor_SongList_Component/evaluation/evaluation.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const os = require('os');
const { execSync } = require('child_process');

function generateRunId() {
  return Math.random().toString(36).substring(2, 10);
}

function getGitInfo() {
  const gitInfo = { git_commit: 'unknown', git_branch: 'unknown' };
  try {
    gitInfo.git_commit = execSync('git rev-parse HEAD', { encoding: 'utf8', timeout: 5000 }).trim().substring(0, 8);
  } catch (err) {}
  try {
    gitInfo.git_branch = execSync('git rev-parse --abbrev-ref HEAD', { encoding: 'utf8', timeout: 5000 }).trim();
  } catch (err) {}
  return gitInfo;
}

function getEnvironmentInfo() {
  const gitInfo = getGitInfo();
  return {
    node_version: process.version,
    platform: os.platform(),
    os: os.type(),
    os_release: os.release(),
    architecture: os.arch(),
    hostname: os.hostname(),
    git_commit: gitInfo.git_commit,
    git_branch: gitInfo.git_branch,
  };
}

function runTests(repositoryPath, repositoryName) {
  console.log('\n' + '='.repeat(60));
  console.log(`RUNNING TESTS: ${repositoryName}`);
  console.log('='.repeat(60));

  try {
    const testCommand = repositoryName === 'repository_before' ? 'npm run test-before' : 'npm run test-after';
    const output = execSync(testCommand, {
      cwd: path.join(__dirname, '..'),
      encoding: 'utf8',
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env, CI: 'true' }
    });

    const lines = output.split('\n');
    const tests = [];
    let passed = 0;
    let failed = 0;
    let total = 0;

    // Parse Jest output for test results
    lines.forEach(line => {
      if (line.includes('✓') || line.includes('PASS')) {
        const testMatch = line.match(/✓\s+(.+)/);
        if (testMatch) {
          tests.push({
            nodeid: `${repositoryName}::${testMatch[1]}`,
            name: testMatch[1],
            outcome: 'passed',
            message: 'Test passed',
          });
          passed++;
        }
      } else if (line.includes('✕') || line.includes('FAIL')) {
        const testMatch = line.match(/✕\s+(.+)/);
        if (testMatch) {
          tests.push({
            nodeid: `${repositoryName}::${testMatch[1]}`,
            name: testMatch[1],
            outcome: 'failed',
            message: 'Test failed',
          });
          failed++;
        }
      }
    });

    // Extract summary from Jest output
    const summaryMatch = output.match(/Tests:\s+(\d+)\s+failed,\s+(\d+)\s+passed,\s+(\d+)\s+total/) ||
                        output.match(/Tests:\s+(\d+)\s+passed,\s+(\d+)\s+total/);
    
    if (summaryMatch) {
      if (summaryMatch.length === 4) {
        failed = parseInt(summaryMatch[1]);
        passed = parseInt(summaryMatch[2]);
        total = parseInt(summaryMatch[3]);
      } else {
        passed = parseInt(summaryMatch[1]);
        total = parseInt(summaryMatch[2]);
        failed = total - passed;
      }
    } else {
      total = passed + failed;
    }

    const summary = {
      total: total,
      passed: passed,
      failed: failed,
      errors: 0,
      skipped: 0,
    };

    console.log(`\n${passed > 0 ? '✅' : '❌'} Tests: ${summary.passed}/${summary.total} passed, ${summary.failed} failed`);

    return {
      success: failed === 0,
      exit_code: failed === 0 ? 0 : 1,
      tests,
      summary,
    };
  } catch (err) {
    console.error('\nERROR:', err.message);
    
    // Parse failed test output
    const output = err.stdout || err.message;
    const lines = output.split('\n');
    let failed = 0;
    let passed = 0;
    let total = 0;
    
    const summaryMatch = output.match(/Tests:\s+(\d+)\s+failed,\s+(\d+)\s+passed,\s+(\d+)\s+total/) ||
                        output.match(/Test Suites:\s+(\d+)\s+failed/);
    
    if (summaryMatch) {
      if (summaryMatch.length === 4) {
        failed = parseInt(summaryMatch[1]);
        passed = parseInt(summaryMatch[2]);
        total = parseInt(summaryMatch[3]);
      } else {
        // Extract from error output
        const failedMatch = output.match(/(\d+)\s+failed/);
        const passedMatch = output.match(/(\d+)\s+passed/);
        const totalMatch = output.match(/(\d+)\s+total/);
        
        failed = failedMatch ? parseInt(failedMatch[1]) : 0;
        passed = passedMatch ? parseInt(passedMatch[1]) : 0;
        total = totalMatch ? parseInt(totalMatch[1]) : failed + passed;
      }
    }

    return {
      success: false,
      exit_code: 1,
      tests: [],
      summary: { total: total, passed: passed, failed: failed, errors: 0, skipped: 0 },
      error: err.message,
    };
  }
}

function generateOutputPath() {
  const now = new Date();
  const dateStr = now.toISOString().split('T')[0];
  const timeStr = now.toTimeString().split(' ')[0].replace(/:/g, '-');
  const outputDir = path.join(__dirname, dateStr, timeStr);
  fs.mkdirSync(outputDir, { recursive: true });
  return path.join(outputDir, 'report.json');
}

const runId = generateRunId();
const startedAt = new Date();

console.log('\n' + '='.repeat(60));
console.log('SONGLIST COMPONENT REFACTOR EVALUATION');
console.log('='.repeat(60));
console.log(`Run ID: ${runId}`);
console.log(`Started at: ${startedAt.toISOString()}`);

const beforeResults = runTests('repository_before', 'repository_before');
const afterResults = runTests('repository_after', 'repository_after');

const finishedAt = new Date();
const duration = (finishedAt - startedAt) / 1000;

const comparison = {
  before_tests_passed: beforeResults.success,
  before_total: beforeResults.summary.total,
  before_passed: beforeResults.summary.passed,
  before_failed: beforeResults.summary.failed,
  after_tests_passed: afterResults.success,
  after_total: afterResults.summary.total,
  after_passed: afterResults.summary.passed,
  after_failed: afterResults.summary.failed,
  improvement: afterResults.summary.passed - beforeResults.summary.passed,
  improvement_percentage: beforeResults.summary.total > 0 ? 
    Math.round(((afterResults.summary.passed - beforeResults.summary.passed) / beforeResults.summary.total) * 100) : 0
};

console.log('\n' + '='.repeat(60));
console.log('EVALUATION SUMMARY');
console.log('='.repeat(60));
console.log(`\nBefore Implementation (repository_before):`);
console.log(`  Overall: ${beforeResults.success ? '✅ PASSED' : '❌ FAILED'}`);
console.log(`  Tests: ${comparison.before_passed}/${comparison.before_total} passed (${Math.round((comparison.before_passed/comparison.before_total)*100)}%)`);
console.log(`\nAfter Implementation (repository_after):`);
console.log(`  Overall: ${afterResults.success ? '✅ PASSED' : '❌ FAILED'}`);
console.log(`  Tests: ${comparison.after_passed}/${comparison.after_total} passed (${Math.round((comparison.after_passed/comparison.after_total)*100)}%)`);
console.log(`\nImprovement:`);
console.log(`  Additional tests passing: +${comparison.improvement}`);
console.log(`  Improvement percentage: +${comparison.improvement_percentage}%`);

const success = afterResults.success && (comparison.improvement > 0 || comparison.after_passed === comparison.after_total);
const errorMessage = success ? null : 'Implementation did not meet success criteria';

const report = {
  run_id: runId,
  started_at: startedAt.toISOString(),
  finished_at: finishedAt.toISOString(),
  duration_seconds: parseFloat(duration.toFixed(6)),
  success,
  error: errorMessage,
  environment: getEnvironmentInfo(),
  results: {
    before: beforeResults,
    after: afterResults,
    comparison,
  },
};

const outputPath = generateOutputPath();
fs.writeFileSync(outputPath, JSON.stringify(report, null, 2));

console.log(`\n✅ Report saved to: ${outputPath}`);
console.log('\n' + '='.repeat(60));
console.log('EVALUATION COMPLETE');
console.log('='.repeat(60));
console.log(`Duration: ${duration.toFixed(2)}s`);
console.log(`Success: ${success ? '✅ YES' : '❌ NO'}`);

process.exit(success ? 0 : 1);

File: cd41ba-Async_circular_data_processing_with_type_safety/evaluation/evaluation.py
#!/usr/bin/env python3
"""
Evaluation runner for Async Circular Data Processing (score).

This evaluation script:
- Runs pytest tests on the tests/ folder for both before and after implementations
- Collects individual test results with pass/fail status
- Generates structured reports with environment metadata

Run with:
    docker compose run --rm app python evaluation/evaluation.py [options]
"""
import sys
import json
import uuid
import platform
import subprocess
from pathlib import Path
from datetime import datetime

ROOT = Path(__file__).resolve().parent.parent
REPORTS = ROOT / "evaluation" / "reports"

def environment_info():
    """Collect environment metadata as requested."""
    info = {
        "python_version": platform.python_version(),
        "platform": platform.platform(),
        "os": platform.system(),
        "os_release": platform.release(),
        "architecture": platform.machine(),
        "hostname": platform.node(),
        "git_commit": "unknown",
        "git_branch": "unknown"
    }
    
    # Try to get git info if available
    try:
        commit = subprocess.check_output(["git", "rev-parse", "HEAD"], stderr=subprocess.DEVNULL).decode().strip()
        branch = subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"], stderr=subprocess.DEVNULL).decode().strip()
        info["git_commit"] = commit
        info["git_branch"] = branch
    except Exception:
        pass
        
    return info

def print_header(title):
    print(f"\n{'=' * 60}")
    print(title)
    print(f"{'=' * 60}")

def parse_pytest_verbose_output(output):
    """
    Parse pytest verbose output to extract test results.
    Returns a list of dicts: {nodeid, name, outcome}
    """
    tests = []
    lines = output.split('\n')
    
    for line in lines:
        line = line.strip()
        if not line:
            continue
        
        # Split line by whitespace to find status
        parts = line.split()
        if len(parts) < 2:
            continue

        # Check for status keyword
        status = None
        status_index = -1
        
        # Valid statuses
        valid_statuses = {"PASSED", "FAILED", "SKIPPED", "ERROR", "XFAIL", "XPASS"}
        
        # Iterate backwards to find status 
        for i, part in enumerate(reversed(parts)):
            if part in valid_statuses:
                status = part.lower()
                status_index = len(parts) - 1 - i
                break
        
        if status:
             # Everything before status is the nodeid
            nodeid = " ".join(parts[:status_index])
            
            # It should look like a test path
            if nodeid.startswith("tests/"):
                # Extract name
                if "::" in nodeid:
                    name = nodeid.split("::")[-1]
                else:
                    # Fallback for file-level errors or flat output
                    name = nodeid.split("/")[-1]

                tests.append({
                    "nodeid": nodeid,
                    "name": name,
                    "outcome": status
                })
    
    return tests

def run_tests_with_reporting(repo_name: str, label: str):
    """
    Run pytest and return detailed result dict matching schema.
    """
    repo_path = ROOT / repo_name
    print_header(f"RUNNING TESTS: {label}")
    print(f"PYTHONPATH: {repo_path}")
    print(f"Tests directory: {ROOT / 'tests'}")
    
    # We pass the repo path as PYTHONPATH
    env = sys.modules['os'].environ.copy()
    existing_path = env.get("PYTHONPATH", "")
    env["PYTHONPATH"] = str(repo_path) + (f"{sys.modules['os'].pathsep}{existing_path}" if existing_path else "")
    
    try:
        proc = subprocess.run(
            [sys.executable, "-m", "pytest", "tests", "-v", "--no-header", "--color=no"],
            cwd=ROOT,
            capture_output=True,
            text=True,
            env=env,
            timeout=120
        )
        
        stdout = proc.stdout
        stderr = proc.stderr
        output = stdout + stderr
        
        # Parse results
        tests = parse_pytest_verbose_output(output)
        
        passed_count = sum(1 for t in tests if t['outcome'] == 'passed')
        failed_count = sum(1 for t in tests if t['outcome'] == 'failed')
        error_count = sum(1 for t in tests if t['outcome'] == 'error')
        skipped_count = sum(1 for t in tests if t['outcome'] == 'skipped')
        total_count = len(tests)
        
        # If parsing failed but we have output, maybe something went wrong with the pattern matching.
        # Print raw output to help debug if running interactively or ensure it's in the log.
        if total_count == 0 and len(output) > 0:
            print("WARNING: No tests parsed from pytest output. Raw output start:")
            print(output[:500])
            print("...")
        
        # Print to stdout as requested
        print(f"\nResults: {passed_count} passed, {failed_count} failed, {error_count} errors, {skipped_count} skipped (total: {total_count})")
        
        for test in tests:
            icon = "✅" if test['outcome'] == 'passed' else "❌" if test['outcome'] == 'failed' else "❓"
            print(f"  {icon} {test['nodeid']}: {test['outcome']}")
        
        return {
            "success": proc.returncode == 0,
            "exit_code": proc.returncode,
            "tests": tests,
            "summary": {
                "total": total_count,
                "passed": passed_count,
                "failed": failed_count,
                "errors": error_count,
                "skipped": skipped_count
            },
            "stdout": stdout,
            "stderr": stderr
        }
    except subprocess.TimeoutExpired:
        print("\n\n❌ TIMEOUT")
        return {
            "success": False,
            "exit_code": -1,
            "tests": [],
            "summary": {"total": 0, "passed": 0, "failed": 0, "errors": 0, "skipped": 0},
            "stdout": "pytest timeout",
            "stderr": ""
        }

def run_evaluation():
    run_id = str(uuid.uuid4())[:8]
    start = datetime.utcnow()
    
    print(f"Run ID: {run_id}")
    print(f"Started at: {start.isoformat()}Z")
    
    print_header("ASNYC CIRCULAR DATA PROCESSING EVALUATION")
    
    before_label = "BEFORE (REPOSITORY_BEFORE)"
    before_result = run_tests_with_reporting("repository_before", before_label)
    
    after_label = "AFTER (REPOSITORY_AFTER)"
    after_result = run_tests_with_reporting("repository_after", after_label)
    
    end = datetime.utcnow()
    duration = (end - start).total_seconds()
    
    # Comparison Logic
    comparison = {
        "before_tests_passed": before_result["success"],
        "after_tests_passed": after_result["success"],
        "before_total": before_result["summary"]["total"],
        "before_passed": before_result["summary"]["passed"],
        "before_failed": before_result["summary"]["failed"],
        "after_total": after_result["summary"]["total"],
        "after_passed": after_result["summary"]["passed"],
        "after_failed": after_result["summary"]["failed"]
    }
    
    # Success definition is After must pass
    success = after_result["success"]
    
    # Print Summary to Stdout
    print_header("EVALUATION SUMMARY")
    print(f"\nBefore Implementation (repository_before):")
    print(f"  Overall: {'✅ PASSED' if before_result['success'] else '❌ FAILED'}")
    print(f"  Tests: {before_result['summary']['passed']}/{before_result['summary']['total']} passed")

    print(f"\nAfter Implementation (repository_after):")
    print(f"  Overall: {'✅ PASSED' if after_result['success'] else '❌ FAILED'}")
    print(f"  Tests: {after_result['summary']['passed']}/{after_result['summary']['total']} passed")
    
    print_header("EXPECTED BEHAVIOR CHECK")
    if success:
        print("✅ After implementation: All tests passed (expected)")
    else:
        print("❌ After implementation: Some tests failed")

    # Construct final report
    report = {
        "run_id": run_id,
        "started_at": start.isoformat(),
        "finished_at": end.isoformat(),
        "duration_seconds": duration,
        "success": success,
        "error": None,
        "environment": environment_info(),
        "results": {
            "before": before_result,
            "after": after_result,
            "comparison": comparison
        }
    }
    
    return report

def main():
    REPORTS.mkdir(parents=True, exist_ok=True)
    report = run_evaluation()
    
    # Generate path based on date/time structure
    now = datetime.utcnow()
    date_str = now.strftime("%Y-%m-%d")
    time_str = now.strftime("%H-%M-%S")
    path = REPORTS.parent / date_str / time_str / "report.json"
    path.parent.mkdir(parents=True, exist_ok=True)
    
    path.write_text(json.dumps(report, indent=2))
    print(f"\n✅ Report saved to: {path}")
    
    print_header("EVALUATION COMPLETE")
    print(f"Run ID: {report['run_id']}")
    print(f"Duration: {report['duration_seconds']:.2f}s")
    print(f"Success: {'✅ YES' if report['success'] else '❌ NO'}")
    
    return 0 if report["success"] else 1

if __name__ == "__main__":
    sys.exit(main())
